compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -13676,7 +13676,9 @@ OMR::Z::TreeEvaluator::primitiveArraycopyEvaluator(TR::Node* node, TR::CodeGener
       TR::Register *checkBoundReg = srm->findOrCreateScratchRegister();
       if (byteLenReg == NULL)
          {
-         byteLenReg = cg->gprClobberEvaluate(byteLenNode);
+         TR_ASSERT_FATAL(isConstantByteLen, "byteLenNode can be not evaluated only when we have constant length");
+         byteLenReg = cg->allocateRegister();
+         genLoadLongConstant(cg, node, byteLenNode->getConst<int64_t>(), byteLenReg);
```

Comment:
I agree, this should be 32-bit `getConst` load.

Reply:
Actually length child of **arraycopy** is `long (64-Bit)`. Extracting important part of the arraycopy trees here. 
```
n8n         arraycopy  java/lang/System.arraycopy(Ljava/lang/Object;ILjava/lang/Object;II)V
n34n          aladd (internalPtr sharedMemory ) ---> srcPt
                ...
n38n          aladd (internalPtr sharedMemory ) ---> dstPt
                ...
n40n          lconst 480 (highWordZero X!=0 X>=0 cannotOverflow ) ---> Length (Bytes)
```
As System.arrayCopy takes length argument as integer, actual number of bytes will be adjusted as per type of array copy so for example for integer array copy it will be `length x 4`. 
So I think after inlining the System.arrayCopy call in trees it adjust the length node to represent number of bytes to be copied.

Reply:
@andrewcraik can you confirm this because I see at least one location where we explicitly make the type a 32-bit integer via `l2i`, so shouldn't the constant also have type `Int32`?

Reply:
I've seen that too.  LoopReducer also creates arraycopies with integer lengths, and I don't see any obvious guards for 32-bit mode only.

Reply:
Thanks. It sounds like then we need to take a look at the x codegen and do something similar. I suspect we have a platform check and extract the right width, so we should be doing the same here.

Reply:
Call `genLoadLongConstant` made here is intentional to make sure that we write whole 64-Bit of the register. We have a helper function that loads 32-Bit immediate value to register that writes in to only lower 4 bytes of the register keeping upper 4 bytes unchanged. As `byteLenReg` is later used to generate a memory reference for vector instruction which is used to accelerate backward array copy this ensures that we do not have a garbage value index register of the memory reference. 
@0xdaryl  just to clarify, you are talking about use of long constant in evaluator or `byteLenNode->getConst<int64_t>()` call?

Reply:
My understanding was that the length node of the arraycopy was an int32.  If that is the case then I would expect you to be asking for an int32_t constant from the node.

While I believe @andrewcraik's explanation is true, I think I need to satisfy myself that the type of the length child we put on the node differs between 32-bit and 64-bit because I *believe* I've seen only an int type used regardless of native bitness.

Reply:
So if the value is not a long / treated as a long on 64bit there is a risk that we could transform a System.arraycopy call to an arraycopy node and us to then do the wrong thing. I did confirm with @vijaysun-omr and the int32 vs int64 distinction was his understanding as well. If we aren't generating the right width then that is a bug where the arraycopy is being generated IMO and one that needs to be fixed.

Reply:
Rebased the branch. @0xdaryl to clarify, On Z we do not have 31-Bit OS, we just have 31 Bit JVM running on 64 Bit OS, so in https://github.com/eclipse/omr/commit/f1a7ebfe68f664f3e1faaf572cabe3d4425fdb31 we folded `use64BitRegsOn32Bit` to always `true` on `31 Bit JVM` to use full 64 Bit register for longs instead of `RegisterPair`. 
Also looking into the code where we set `_constValue` of `_unionBase` on node, I see even when we pass `int32_t`  , it casts value to `int64_t` and stores. so Getting a 64 Bit value on given node should be fine.